flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/JdbcLockFactory.java

```diff
@@ -0,0 +1,246 @@
+      throw new UncheckedSQLException(
+          e, "Cannot initialize JDBC table maintenance lock: Connection failed");
+    } catch (SQLException e) {
+      throw new UncheckedSQLException(e, "Cannot initialize JDBC table maintenance lock");
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new UncheckedInterruptedException(e, "Interrupted in call to initialize");
+    }
+  }
+
+  public static class Lock implements TriggerLockFactory.Lock {
+    private final JdbcClientPool pool;
+    private final String issuer;
+    private final Type type;
+
+    public Lock(JdbcClientPool pool, String issuer, Type type) {
+      this.pool = pool;
+      this.issuer = issuer;
+      this.type = type;
+    }
+
+    @Override
+    public boolean tryLock() {
+      if (isHeld()) {
+        LOG.info("Lock is already held");
+        return false;
+      }
+
+      try {
+        return pool.run(
+            conn -> {
+              try (PreparedStatement sql = conn.prepareStatement(CREATE_LOCK_SQL)) {
+                sql.setString(1, type.key);
+                sql.setString(2, issuer);
+                sql.setString(3, UUID.randomUUID().toString());
+                return sql.executeUpdate() == 1;
+              }
+            });
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new UncheckedInterruptedException(e, "Interrupted during tryLock");
+      } catch (SQLException e) {
+        // SQL exception happened when creating the lock
+        throw new UncheckedSQLException(e, "Failed to create %s lock", type);
+      }
+    }
+
+    @SuppressWarnings("checkstyle:NestedTryDepth")
+    @Override
+    public boolean isHeld() {
+      try {
+        return pool.run(
+            conn -> {
+              try (PreparedStatement sql = conn.prepareStatement(GET_LOCK_SQL)) {
+                sql.setString(1, type.key);
+                sql.setString(2, issuer);
+                try (ResultSet rs = sql.executeQuery()) {
+                  return rs.next();
+                }
+              }
+            });
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new UncheckedInterruptedException(e, "Interrupted during isHeld");
+      } catch (SQLException e) {
+        // SQL exception happened when getting lock information
+        throw new UncheckedSQLException(e, "Failed to get lock information for %s", type);
+      }
+    }
+
+    @SuppressWarnings("checkstyle:NestedTryDepth")
+    @Override
+    public void unlock() {
+      try {
+        pool.run(
+            conn -> {
+              String lockId;
+              try (PreparedStatement sql = conn.prepareStatement(GET_LOCK_SQL)) {
+                sql.setString(1, type.key);
+                sql.setString(2, issuer);
+                try (ResultSet rs = sql.executeQuery()) {
+                  if (rs.next()) {
+                    lockId = rs.getString(1);
```

Comment:
The `tryLock` and the `unlock` will happen on a different object instance, possibly on a different JVM even. So that doesn't help. We need to fetch the id in the `unlock` method